#[187364735] users profiles endpoints

Yvan-David opened the pull request:
#### What does this PR do?

- implement display of user profile endpoints
- implement user input  profile endpoints
- implement update user profile  profile endpoints
- implement validation on any user access

#### Description of Task to be completed?
 Profiles will provide a way to maintain generic and user specific data/info.

#### How should this be manually tested?

- Screen record (Loom link) : [Testing user Profile end points](https://www.loom.com/share/ef038e31774b414ea868c4e947b83f11?sid=da8dc5e4-f633-4430-8d70-ad30a97e6316)

-  Using swagger documentation
-  Using postman

#### Any background context you want to provide?

#### What are the relevant pivotal tracker/Trello stories?

 Profile Page Settings:  #187364735

#### Screenshots (if appropriate)

#### Questions:

munezerobagira commented on .codeclimate.yml:
```diff
@@ -2,11 +2,20 @@ version: '2'
 checks:
   complex-logic:
     enabled: false
+  method-complexity:
+    enabled: false
   method-lines:
     config:
       threshold: 50
     eslint:
       enabled: true
+  similar-code:
+    config:
+      threshold: 75
+  identical-code:
+    config:
+      threshold: 75
+
```
why did change this configurations?

munezerobagira commented on package.json:
```diff
@@ -5,10 +5,13 @@
+    "dev": "nodemon -L --exec ts-node index.ts --watch src --ext ts,js",
+    "build": "tsc && npm run migrate && npm run seed",
+    "test": "cross-env NODE_ENV=test && tsc && npm run migrate && nyc mocha --require ts-node/register test/*.ts",
+    "coverage": "cross-env NODE_ENV=test nyc --reporter=text --reporter=lcov mocha --require ts-node/register test/*.ts --coverage && nyc report --reporter=text-lcov | coveralls",
```
nice

Yvan-David replied:
Thanks

munezerobagira commented on src/controllers/userProfileController.ts:
```diff
@@ -0,0 +1,42 @@
+  try {
+    const { id } = req.params
+    const fieldsToUpdate = req.body
+    const { error } = profileValidationalSchema.validate(fieldsToUpdate)
```
```suggestion
```
you can move this validation to middleware

munezerobagira commented on src/database/migrations/20240430000533-add-profile-fields.js:
you can create new table for profile only, and make sure that the user profile is created upon login and add one to one relationship

Yvan-David replied:
Done

munezerobagira commented on src/database/models/userModel.ts:
```diff
@@ -25,6 +53,45 @@ class User extends Model<UserAttributes> implements UserAttributes {
   public password!: string
 
   public userRole!: string
+
+  public gender!: string
+
+  public birthdate!: string
+
+  public phoneNumber!: string
+
+  public preferredLanguage!: string
+
+  public preferredCurrency!: string
+
+  public homeAddress!: string
+
+  public billingAddress!: string
+
+  public profileImageUrl?: string
+
+  public readonly createdAt!: Date
+
+  public readonly updatedAt!: Date
+
```
```suggestion

```
it would be better if you had user profile as new table and add one to one relationship

Yvan-David replied:
Done

munezerobagira commented on src/controllers/userProfileController.ts:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    const user = await User.findOne({ where: { id } })
+    const result = await User.update(fieldsToUpdate, {
```
```suggestion
```
remember to add proper check for updating email or password, updating email requires email verification

munezerobagira reviewed (changes requested):
Good work but there few things which we can improve

EricNiyo73 reviewed (commented):
Try to show us how this feature works , by providing the valid loom link.

Yvan-David commented:
> Try to show us how this feature works , by providing the valid loom link.

Done

EricNiyo73 reviewed (commented):
Update your swagger docs and use form-data ,In order to shows us clearly how you save Imageurl and how it is saved well into cloudinary as I have seen from this PR codes.

Yvan-David commented:
> Update your swagger docs and use form-data ,In order to shows us clearly how you save Imageurl and how it is saved well into cloudinary as I have seen from this PR codes.

Done Thanks

EricNiyo73 reviewed (commented):
fix this coveralls checks

![image](https://github.com/atlp-rwanda/e-commerce-bitcrafters-bn/assets/106761137/b2f7e13b-9df7-4b2b-a269-783be1d16849)

munezerobagira commented on src/middlewares/validate.ts:
```diff
@@ -0,0 +1,25 @@
+  (schema: Joi.ObjectSchema, type: requestType = requestType.body) =>
+  (req: Request, res: Response, next: NextFunction) => {
+    try {
+      const { error } = schema.validate(req.body, { abortEarly: false });
```
```suggestion
      const { error } = schema.validate(type, { abortEarly: false });

```

munezerobagira commented on src/controllers/userProfileController.ts:
These userProfile controller functions can moved to UserController as they are related

munezerobagira commented on src/database/models/userModel.ts:
```diff
@@ -25,7 +26,12 @@ class User extends Model<UserAttributes> implements UserAttributes {
+    User.hasOne(UserProfile, {
+      foreignKey: 'userId'
+    });
+  }}
```
Good work for adding relationships

munezerobagira replied:
but We can also add other constraints like onDelete cascade, to make sure that the profile is deleted when the user is deleted,

munezerobagira commented on src/helpers/swagger.ts:
```diff
@@ -1,6 +1,6 @@
 import swaggerJsdoc, { Options } from 'swagger-jsdoc'
 
-const port = process.env.PORT || 8000 // Assuming you have PORT defined somewhere
+const port = process.env.PORT || 3000 // Assuming you have PORT defined somewhere
```
```suggestion

```
why would you change  default PORT value while it can be changed through env?

munezerobagira commented on src/middlewares/authentication.ts:
```diff
@@ -0,0 +1,23 @@
+import { NextFunction, Request, Response } from 'express'
+import jwt, { JwtPayload } from 'jsonwebtoken'
+
+const SECRET = process.env.JWT_SECRET
```
```suggestion
import  {JWT_SECRET}  from "../config/constants"
```
If possible, we can move this into constants for easier access on  environment variables

munezerobagira commented on src/database/models/userProfile.ts:
```diff
@@ -0,0 +1,137 @@
+
+  public readonly updatedAt!: Date
+  
+  public static associate() {
+    UserProfile.belongsTo(User)
+   }
```
Nice

munezerobagira commented on src/database/models/userProfile.ts:
```diff
@@ -0,0 +1,137 @@
+      primaryKey: true,
+    },
+    userId: {
+      type: DataTypes.INTEGER,
```
```suggestion
      type: DataTypes.INTEGER,
      allowNull: false,
```

munezerobagira reviewed (changes requested):
Good work but there are things that we can improve

Furaha-Justine reviewed (commented):
Rebase to resolve conflicts.

AngeloChristian1 committed:
feat(user profile endPoints): implement update and get userProfile
- create user profile table and link it with user table
- implement get endpoint
- implement update endpoint

[Delivers #187364735]

AngeloChristian1 commented on src/controllers/UserController.ts:
```diff
@@ -33,4 +35,44 @@ export default class UserController {
+        await User.update({username, email}, {
+        where: { id },
+      })
+      await UserProfile.update(fieldsToUpdate, {
+        where: { userId: id },
+      })
+  
+      const updateduser = await UserProfile.findOne({
```
can you user services for these

AngeloChristian1 commented on src/controllers/UserController.ts:
```diff
@@ -33,4 +35,44 @@ export default class UserController {
+  static async getUser(req: Request, res: Response): Promise<Response> {
+    try {
+      const { id } = req.params
+      const myprofile = await UserProfile.findOne({
```
here also you can user getUserById service in src/services/userServices.ts

AngeloChristian1 commented on src/database/models/userProfile.ts:
```diff
@@ -0,0 +1,138 @@
+export interface UserProfileAttributes {
+  id: number
+  userId: number
+  username: string
+  email: string
+  birthdate: string
+  gender: string
+  preferredLanguage: string
+  preferredCurrency: string
+  phoneNumber: string
+  homeAddress: string
+  billingAddress: string
+  profileImageUrl: string
+  createdAt: Date
+  updatedAt: Date
```
You can also make these optional, cause sometimes a user may not have filled out all out the

AngeloChristian1 commented on src/middlewares/authenticationMiddleware.ts:
```diff
@@ -35,5 +37,22 @@ const isAuthenticated = async (
   }
 }
-
+export const isOwner = async (
+  req: Request,
+  res: Response,
+  next: NextFunction,
+) => {
+  try {
+    const { id } = req.params
+    const authHeader = req.headers.authorization
+    const token = authHeader.split(' ')[1]
+    const decoded = jwt.verify(token, JWT_SECRET) as JwtPayload & { id: string }
```
This part is already covered by the isAuthenticated middleware, you can put this as a secondary middleware and start by reading the `user` from the request

EricNiyo73 replied:
This middleware it is not needed, it seems like you are duplicating the codes.

EricNiyo73 replied:
This middleware it is not needed, it seems like you are duplicating the codes.
Just try resolve that.

AngeloChristian1 commented on src/routes/userRoute.ts:
```diff
@@ -1,14 +1,23 @@
   validateRequest(userSchema, 'body'),
   UserController.signup,
 )
+router.get(
+  '/profile/:id',isOwner, UserController.getUser)
+
+router.patch(
+  '/profile/:id',isOwner, singleFileUpload,
+  validateRequest(profileValidationalSchema, 'body'),
+  UserController.updateUser
+)
```
you can fist pass in isAuthenticated, so that it can decode the authorisation token and pass the decoded in the, 
request
```suggestion
router.get(
  '/profile/:id',isAuthenticated,isOwner, UserController.getUser)

router.patch(
  '/profile/:id',isAuthenticated,isOwner, singleFileUpload,
  validateRequest(profileValidationalSchema, 'body'),
  UserController.updateUser
)
```

EricNiyo73 replied:
Here, It seems like here you tried to pass two middleware's which is not necessary , may be you can pass only isAuthenticated middleware, because I have seen that it decodes also the token and also checks if that user is known in our database and then finally set the user into the request.

We don't need to take id in params, we can take all information from the token.

EricNiyo73 replied:
Here, It seems like here you tried to pass two middleware's which is not necessary , may be you can pass only isAuthenticated middleware, because I have seen that it decodes also the token and also checks if that user is known in our database and then finally set the user into the request.

We don't need to take id in params, we can take all information from the token.

AngeloChristian1 commented on src/utils/cloudinary.ts:
good job

AngeloChristian1 reviewed (commented):
Good job, we can work on improvements

EricNiyo73 reviewed (changes requested):
Try to resolve that issues of middlewares

munezerobagira reviewed (approved):
Nice
